File: nildb/secretvault_python/generate.py
import jwt
import time
from ecdsa import SigningKey, SECP256k1
from config import NODE_CONFIG, ORG_DID, ORG_SECRET_KEY

def create_jwt(secret_key: str = None,
               org_did: str = None,
               node_ids: list = None,
               ttl: int = 3600) -> list:
    """
    Create JWTs signed with ES256K for multiple node_ids
    """
    
    # Convert the secret key from hex to bytes
    private_key = bytes.fromhex(secret_key)
    signer = SigningKey.from_string(private_key, curve=SECP256k1)

    tokens = []
    for node_id in node_ids:
        # Create payload for each node_id
        payload = {
            "iss": org_did,
            "aud": node_id,
            "exp": int(time.time()) + ttl
        }

        # Create and sign the JWT
        token = jwt.encode(
            payload,
            signer.to_pem(),
            algorithm="ES256K"
        )
        tokens.append(token)
        print(f"Generated JWT for {node_id}: {token}")
    
    return tokens

if __name__ == "__main__":
    secret_key = ORG_SECRET_KEY
    org_did = ORG_DID
    node_ids = [node['url'] for node in NODE_CONFIG.values()]
    create_jwt(secret_key, org_did, node_ids)

File: nildb/secretvault_nextjs_nilql/generate.js
// generate.js
// npm install did-jwt
// run it via node (node generate.js)

const { createJWT, ES256KSigner } = require('did-jwt');
const { Buffer } = require('buffer');

/**
 * Create JWTs signed with ES256K for multiple node_ids
 * @param {string} secretKey - Hex string of the private key
 * @param {string} orgDid - Organization DID
 * @param {string[]} nodeIds - Array of node DIDs
 * @param {number} ttl - Time to live in seconds
 * @returns {Promise<string[]>} Array of JWT tokens
 */
async function createJwt(secretKey, orgDid, nodeIds, ttl = 3600) {
  // Create signer from private key
  const signer = ES256KSigner(Buffer.from(secretKey, 'hex'));
  const tokens = [];

  for (const nodeId of nodeIds) {
    const payload = {
      iss: orgDid,
      aud: nodeId,
      exp: Math.floor(Date.now() / 1000) + ttl,
    };

    const token = await createJWT(payload, { issuer: orgDid, signer });
    tokens.push(token);
    console.log(`Generated JWT for ${nodeId}: ${token}`);
  }

  return tokens;
}

// Example usage
// Replace SECRET_KEY and ORG_DID with your own values
async function main() {
  const secretKey = 'did:nil:testnet:nillionXXX';
  const orgDid = 'XXX';
  const nodeIds = [
    'did:nil:testnet:nillion1qfrl8nje3nvwh6cryj63mz2y6gsdptvn07nx8v',
    'did:nil:testnet:nillion1uak7fgsp69kzfhdd6lfqv69fnzh3lprg2mp3mx',
    'did:nil:testnet:nillion1kfremrp2mryxrynx66etjl8s7wazxc3rssrugk',
  ];

  await createJwt(secretKey, orgDid, nodeIds);
}

if (require.main === module) {
  main().catch(console.error);
}

module.exports = { createJwt };
